crates/hir/src/code_model.rs

```diff
@@ -724,8 +743,43 @@ impl Function {
     pub fn diagnostics(self, db: &dyn HirDatabase, sink: &mut DiagnosticSink) {
         hir_ty::diagnostics::validate_body(db, self.id.into(), sink)
     }
+
+    /// Whether this function declaration has a definition.
+    ///
+    /// This is false in the case of required (not provided) trait methods.
+    pub fn has_body(self, db: &dyn HirDatabase) -> bool {
+        db.function_data(self.id).has_body
+    }
+
+    /// If this function is a method, the trait or type where it is declared.
+    pub fn method_owner(self, db: &dyn HirDatabase) -> Option<MethodOwner> {
+        match self.as_assoc_item(db).map(|assoc| assoc.container(db)) {
+            Some(AssocItemContainer::Trait(t)) => Some(t.into()),
+            Some(AssocItemContainer::ImplDef(imp)) => {
+                let resolver = ModuleId::from(imp.module(db)).resolver(db.upcast());
+                let ctx = TyLoweringContext::new(db, &resolver);
+                let adt = Ty::from_hir(
+                    &ctx,
+                    &imp.target_trait(db).unwrap_or_else(|| imp.target_type(db)),
+                )
+                .as_adt()
+                .map(|t| t.0)
+                .unwrap();
+                Some(Adt::from(adt).into())
+            }
+            None => None,
+        }
+    }
 }
 
+#[derive(Debug)]
+pub enum MethodOwner {
+    Trait(Trait),
+    Adt(Adt),
```

Comment:
This should be an `ImplDef`, as inherent impls not necessary refer to adts. And, with this change, I think this becomes equivalent to `AssocItemContainer`?

Reply:
> This should be an ImplDef, as inherent impls not necessary refer to adts

Hence the method instead of function naming. 

> And, with this change, I think this becomes equivalent to AssocItemContainer

The difference is that `AssocItemContainer` holds an `ImplDef`, rather than the type that the `impl` targets. To get to the type you need to use `target_type(...) -> TypeRef` and then use a `TyLoweringContext`.

That logic could potentially be moved to a method on `ImplDef` instead?

Reply:
>That logic could potentially be moved to a method on ImplDef instead?

I don't think this logic belongs to `hir` layer. `hir` should always be precise, only in the `ide` we can cut corners and not consider some edge cases. As it is possible that ImplBlock refers to something which is not an adt, hir should't single out adts.